### PR TITLE
feat(CI/nix): init docs on CI and add pre-commit hooks

### PR DESCRIPTION
**Pull request by DavHau**
### Summary
- Add a .md file for each existing github workflow .yml documenting the workflow
- Introduce [cachix/pre-commit-hooks.nix](https://github.com/cachix/pre-commit-hooks.nix)
- add pre-commit-hook to check the nix formatting
- add pre-commit-hook to ensure that all workflow .yml files are documented
- remove the `Unset variables: ...` prints from the pure shell script writer

@jost-s @steveeJ 

### TODO:
- [x] CHANGELOG(s) updated with appropriate info
- [ ] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading

### CODE REVIEW

File: .github/workflows/autoupdate.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# autorebase.yml
+
+Operates on open PRs and automatically merges updates on target branches into the PR branch
+
+# TODO
+
+We already have `autorebase.yml`. Why is this still needed?
```
Comment (steveeJ) [inline .github/workflows/autoupdate.md:7]: autoupdate creates merge commits while autorebase does a rebase

File: .github/workflows/build.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+## build.yml
+
+Runs holochain tests:
```
Comment (steveeJ) [inline .github/workflows/build.md:3]: i find it too early to document it in such detail, if at all. it requires manual updating of the markdown file every time tests are changed or relocated

File: .github/workflows/check_run_detect_release_pr_failure.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# check_run_detect_release_pr_failure.yml
+
+Intentifies new branches created with `releas-` prefix and warns users about this in mattermost:
```
Comment (steveeJ) [inline .github/workflows/check_run_detect_release_pr_failure.md:3]:
```suggestion
Intentifies new branches created with `release-` prefix and warns users about this in mattermost:
```

**Review by steveeJ — COMMENTED** [review]
on a high level i'm worried about creating maintenance burden with this. specifically, separating the doc file from the workflow file and documenting the workflows in such detail.

i would prefer if we added a header comment in the workflow files directly.

File: .github/workflows/autoupdate.md (hunk 1 above)
Comment (DavHau) [reply .github/workflows/autoupdate.md:7]: Any reason why we need both approaches? They seem to solve the same problem.

Comment (steveeJ) [reply .github/workflows/autoupdate.md:7]: functionally they're different and then it comes down to preference

Comment (jost-s) [reply .github/workflows/autoupdate.md:7]: ?? Aren't both run on every push to develop?
And whose preference, I mean how does one get to choose?

File: .github/workflows/check_run_detect_release_pr_failure.md (hunk 3 above)
Comment (jost-s) [inline .github/workflows/check_run_detect_release_pr_failure.md:3]:
```suggestion
Identifies new branches created with `release-` prefix and warns users about this in mattermost:
```

File: .github/workflows/ssh-session.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# ssh-session.yml
+
+An action triggered manually to debug the cached state for `releas.yml` and `release-prepare.yml`
```
Comment (jost-s) [inline .github/workflows/ssh-session.md:3]:
```suggestion
An action triggered manually to debug the cached state for `release.yml` and `release-prepare.yml`
```

**Review by jost-s — CHANGES_REQUESTED** [review]
I think these additions are helpful. I also prefer the docs to be moved to the workflow files for easier maintenance.

File: .github/workflows/autoupdate.md (hunk 1 above)
Comment (DavHau) [reply .github/workflows/autoupdate.md:7]: Autoupdate specifies to only act if the `autoupdate:opt-in` label is set. Potentially there is another label to activate the autorebase. This should be documented somewhere and potentially be added to the PR template if we want contributors to make use of these.

**Comment by DavHau** [comment]
I moved the docs inside the .yml files as comments and applied the suggested fixes.
Concerning too much detail, I removed some for build.yml but kept it for release.yml and release-prepare.yml. I think for those it's currently harder to get an overview due to their sizes and therefore the bullet points are important.

Does this look fine for you @steveeJ ?

Also could you give ma a hint of why the `check_run_detect_release_pr_failure.yml` is needed, so I can add that to the docs?

File: .github/workflows/autoupdate.md (hunk 1 above)
Comment (steveeJ) [reply .github/workflows/autoupdate.md]: they always run but they only do something if the PR is labelled accordingly. the label names are `auto{rebase,update}:opt-in` respectively, and can be applied to a PR by the author or maintainers. one will rebase source onto target and the other will merge target into source.

**Comment by steveeJ** [comment]
> Also could you give ma a hint of why the `check_run_detect_release_pr_failure.yml` is needed, so I can add that to the docs?

i don't think we need it anymore. it was used to detect failed PR checks that are part of a release. but we opted for pushing to github directly during a release.

**Comment by DavHau** [comment]
I removed the `check_run_detect_release_pr_failure.yml` and added more comments to the autoupdate and autorebase workflows.

I think this is ready.

File: nix/modules/pre-commit.nix [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{ self, lib, ... }: {
+  perSystem = { config, self', inputs', pkgs, ... }:
+    {
+      pre-commit.check.enable = true;
```
Comment (steveeJ) [inline nix/modules/pre-commit.nix:4]: what kind of check is enabled here? is it just the one in the next line?

File: nix/modules/devShells.nix [hunk 6]
```diff
@@ -65,6 +65,7 @@
               export CARGO_CACHE_RUSTC_INFO=1
               export PATH="$CARGO_INSTALL_ROOT/bin:$PATH"
               export NIX_PATH="nixpkgs=${pkgs.path}"
+              ${config.pre-commit.installationScript}
```
Comment (steveeJ) [inline nix/modules/devShells.nix:68]: the holonix devshell inherits from rustDev. i don't know if it's appropriate to make this the default for all holonix users so i suggest we move this to coreDev.

Comment (DavHau) [reply nix/modules/devShells.nix:68]: This makes the pre-commit checks become part of `nix flake check`, see: https://flake.parts/options/pre-commit-hooks-nix.html#opt-perSystem.pre-commit.check.enable

File: nix/modules/pre-commit.nix (hunk 5 above)
Comment (DavHau) [reply nix/modules/pre-commit.nix:4]: This makes the pre-commit checks become part of `nix flake check`, see: https://flake.parts/options/pre-commit-hooks-nix.html#opt-perSystem.pre-commit.check.enable

**Review by steveeJ — APPROVED** [review]

**Review by jost-s — APPROVED** [review]
Good stuff

**Pull request merged**